Http2 inbound flow control

jrudolph opened the pull request:
The tests pass but the code is not yet in its final shape. A few constants need to be converted to configuration values. There's currently only a simplistics strategy of sending out WINDOW_UPDATE frames. It is configured with maximum values for connection-level and stream-level windows. When outstanding + buffered data falls below half the maximum configured number it sends out a WINDOW_UPDATE frame to the peer.

I think there are a few hard flow-control problems that cannot be easily solved (but putting it in strategies should help experimenting with other ones):
 * If `max-stream-level-window * num_connections > max-connection-level-window`, there can be head-of-line blocking between streams. Since this is only for uploads, it doesn't seem to be a serious concern for now, the client can still prioritize streams in any way it likes.
 * If the [BDP](https://en.wikipedia.org/wiki/Bandwidth-delay_product) is lower than the configured connection-level window, then there will be head-of-line blocking also for control frames (stuck in TCP buffers at the peer). If the BDP is higher than the configured connection-level window, then the full bandwidth will not be used. Any better solution would at least need some way of estimating the real, current but dynamic BDP. This problem is probably as hard as any kind of TCP congestion control. One potential building block could be trying to read the current BDP as estimated by the OS (`tcp_info` structure in Linux).

Refs #737

ktoso commented on akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala:
```diff
@@ -77,32 +120,107 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with GenericOut
         maybeFinishStream(d.endStream)
       case r: RstStreamFrame ⇒
-        outlet.fail(new PeerClosedStreamException(r.streamId, r.errorCode))
+        buffer.onRstStreamFrame(r)
```
btw; love the general move to these named on* callbacks (everywhere in the infra) 👍

ktoso commented on akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala:
```diff
@@ -77,32 +120,107 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with GenericOut
+        s"adjusting con-level window by $connectionLevel, stream-level window by $streamLevel," +
+          s"remaining window space now $currentWindowSpace, buffered: ${buffer.size}," +
+          s"remaining connection window space now $outstandingConnectionLevelWindow, total buffered: $totalBufferedData")
+    }
```
very clean, like it a lot

2m commented on akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala:
```diff
@@ -62,11 +109,7 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with GenericOut
-            (Source.fromGraph(subSource.source), Open(new BufferedOutlet[ByteString](subSource) {
-              override def onDownstreamFinish(): Unit =
-                multiplexer.pushControlFrame(RstStreamFrame(streamId, ErrorCode.CANCEL))
-              incomingStreams -= streamId
```
Ooh. There was a bug here, right? Missing parenths.

jrudolph replied:
Yes, without that statement in the `onDownstreamFinish` we would have leaked memory for already closed streams.

jrudolph commented on akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala:
```diff
@@ -77,32 +120,107 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with GenericOut
     }
   }
-  sealed abstract class ReceivingData(outlet: BufferedOutlet[ByteString], afterEndStreamReceived: IncomingStreamState) extends IncomingStreamState { _: Product ⇒
+  class Buffer(streamId: Int, outlet: SubSourceOutlet[ByteString]) extends OutHandler {
```
WDYM?

ktoso replied:
just naming of the class, "Buffer" is a bit bland, though I guess it's only used here hm... still :)

jrudolph replied:
You mean to implement auto-to-strict in HTTP/2?

jrudolph replied:
Ah, I see, yeah that would make sense.

jrudolph replied:
Renamed it and moved it around a bit.

jrudolph commented:
Updated. The biggest change is the introduction of a settings structure. I chose to implement the settings directly in akka-http-core and put them below `ServerSettings`. That allows us to keep `akka-http2-support` as basically implementation only. I added a stub trait `Http2InternalServerSettings` that can later be used to allow customizing internal behavior like changing those strategies.

jonas commented on akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala:
```diff
@@ -115,6 +132,75 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with GenericOut
+      log.debug(
+        s"adjusting con-level window by $connectionLevel, stream-level window by $streamLevel," +
+          s"remaining window space now $outstandingStreamWindow, buffered: ${buffer.size}," +
+          s"remaining connection window space now $outstandingConnectionLevelWindow, total buffered: $totalBufferedData")
```
Bikeshedding (from looking at review comments): missing space after commas when the string is constructed like this.

jrudolph commented:
PLS BUILD

jrudolph commented:
PLS BUILD

jrudolph commented:
Thanks for the reviews!
